core/primitives/src/block.rs

```diff
@@ -152,7 +148,9 @@ impl BlockHeaderInnerRest {
 #[derive(BorshSerialize, BorshDeserialize, Serialize, Debug, Clone, PartialEq, Eq)]
 pub struct Approval {
     pub parent_hash: CryptoHash,
-    pub reference_hash: CryptoHash,
+    pub reference_hash: Option<CryptoHash>,
+    pub target_height: BlockHeight,
+    pub is_endorsement: bool,
```

Comment:
If I understand the code correctly, `Approval` can be an enum, which would express the type in a safer way.